### ISSUE TITLE
IWYU should not clobber `// IWYU pragma` comments

### ISSUE THREAD

**Issue by Jacobfaib** [include-what-you-use/include-what-you-use#1484]
`include-what-you-use` and `fix_includes.py` currently clobber the `IWYU pragma` comments if `-Xiwyu --update_comments` or `--update_comments` is passed. This makes it non-idempotent:
```c++
// iwyu_bug.cpp
#include <vector> // IWYU pragma: keep

int main()
{
  std::vector<int> v;
}
```
```sh
$ include-what-you-use -Xiwyu --update_comments ./iwyu_bug.cpp

fix_includes_bug.cpp should add these lines:

fix_includes_bug.cpp should remove these lines:

The full include-list for fix_includes_bug.cpp:
#include <vector>  // for vector
---
```
If the pragmas are there, it stands to reason their placement is intentional!

**Comment by kimgr** [comment]
Huh. I don't see any explicit handling of `// IWYU pragma` in fix_includes.py. There are actually a few tests with `// IWYU pragma: keep`, I need to double-check what results they actually produce.

**Comment by kimgr** [comment]
OK, so these tests didn't have any `// for xyz` comment on the pragma line, and so didn't cover this scenario. But it seems to me `include-what-you-use` itself should probably not update lines with pragma comments in the first place.

**Comment by Jacobfaib** [comment]
Yeah, BTW the true error here was that iwyu would -- for whatever reason -- remove a header which was actually needed[^1], and so I would manually add the header back and insert an `// IWYU pragma: keep`. Now on the second pass iwyu would suddenly realize the header *was* needed but it would update the comment line from the pragma to `// for xyz`. Finally, on the third pass it would once again change its mind and remove the header again :)

[^1]: I've been desperately trying to come up with a minimal reproducer for this to make a bug report but so far no luck. But in case it jogs your memory, I suspect it's a problem with properly resolving proxy objects. We rolled our own `shared_ptr` and iwyu fails to realize that it does indeed need the full definition of the pointed-to class (not just a forward decl) to resolve `operator->`...

**Comment by kimgr** [comment]
I was following the trail of crumbs through IWYU for pragma comments. It turns out IWYU doesn't keep track of what the original line looked like, it only captures the `<includename>` or `"includename"` (even if they're behind a macro, I think) and then synthesizes a new include line using simply `"#include " + quoted_include`.

It's not at all clear to me how that doesn't cause more havoc, but it also means it's a bit more work to fix this in the `--update_comments` case.

I think longer-term I'd like to preserve the include line verbatim, and maybe heuristically rewrite comments with some special cases for pragmas. Or remember that a line was protected by a keep pragma and re-generate it when printing the line back, overriding any `// for`-comments. But this turned out to be a bigger change than I can muster tonight :)

**Comment by Jacobfaib** [comment]
No worries. We have found a workaround locally (just disable the comments), so are more than happy to wait for this to be resolved!
